Refactor shell code in testsuite

praveenkumar opened the pull request:
- Create shell instance only for the command execute
- Remove channel dependency to get the exitCode
- Use go capability to get exitcode for a process instead running a custom command
- Refactor shell struct configuration

praveenkumar committed:
Refactor shell code in testsuite

- Create shell instance only for the command execute
- Remove channel dependency to get the exitCode
- Use go capability to get exitcode for a process instead running a custom command
- Refactor shell struct configuration

Current implementation start the shell as as backend process and then every command
which is pass to Execute function run it on that backend process. This backend shell
process connected to stdPipe/inPipe and errPipe, so the command execution process is
something following.

ShellInstance() => Execute(<cmd>) => shellProcess.inPipe(<cmd>) => Use goroutine to read std/stderr
=> Channel to communicate the exitCode of the command from that goroutine. => ShellCleanup()

What was happening in above process that end to end there is only one shell instance
and if some command hang during the execution from the inPipe it will block rest of
execution.

With this patch following is expected.
Execute(<cmd>) => ShellInstance => Run the command => ShellCleanup()

We now create shell instance, put the command as stdin and
cmd.Run() take care of cleanup of that shell instance.
This way if command hang then it doesn't block the next execution.

cfergeau commented:
>     * Create shell instance only for the command execute
> 
>     * Remove channel dependency to get the exitCode
> 
>     * Use go capability to get exitcode for a process instead running a custom command
> 
>     * Refactor shell struct configuration

Reading the PR description, I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?

guillaumerose commented:
Why do we need a shell after all ? Can't we just run the command ?

praveenkumar commented:
> Why do we need a shell after all ? Can't we just run the command ?

@guillaumerose When you say run the command does that mean directly `exec.Command()` or some other way?

praveenkumar committed:
Add scenario for command not found

guillaumerose commented:
Yes, `exec.Command("crc", "setup")`

praveenkumar committed:
Add platform tag for scenario

This will help to run platform specific test by providing the platform
name as part of godog.tags

praveenkumar committed:
Add windows scenario for test

praveenkumar commented:
> Yes, exec.Command("crc", "setup")

In that case we need to change all our integration test definition where we use something like `executing "sudo podman build -t hello:test -f Dockerfile" succeeds`, in the background we take whatever part of `" <> "` and pass it to shell so we don't need to parse the name and command args.

praveenkumar committed:
Appveyor: add godog tag to windows

guillaumerose commented:
We can do it automatically maybe? Parse the string and use the first element as cmd but...

In CRC there is this kind of test:
```
        When executing "POD=$(oc get pod -o jsonpath="{.items[0].metadata.name}")" succeeds
        And executing "echo $POD" succeeds
        And executing "oc delete pod $POD --now" succeeds
```

Impossible to do without a shell, so let's keep the shell until we have ginkgo :)

praveenkumar commented:
> I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?

@cfergeau These are not exactly different subset of change but a refactor around shell which I am not able to do in different commit messages without breaking a previous commit. What I can do is add another function just to do what `Execute` now doing => delete old suff and rename that function to `Execute`. I am not able to break this tightly coupled commit.

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +88,62 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
+	go func() {
+		defer inPipe.Close()
+		io.WriteString(inPipe, command+"\n")
+	}()
```
`cmd.Stdin =  strings.NewReader(command+"\n")` might work here?

cfergeau commented on testsuite/shell.go:
```diff
@@ -17,53 +17,34 @@ limitations under the License.
-
-	outScanner *bufio.Scanner
-	errScanner *bufio.Scanner
+	checkExitCodeCmd int
```
This should be renamed.

cfergeau commented on testsuite/shell.go:
```diff
@@ -17,53 +17,34 @@ limitations under the License.
 var (
-	shell ShellInstance
+	exitError *exec.ExitError
+	pathError *os.PathError
```
These do not need to be global variables.

cfergeau commented:
> > I was expecting several commits to cover these to make it easier to understand which hunk is related to which of these 4 changes?
> 
> @cfergeau These are not exactly different subset of change but a refactor around shell which I am not able to do in different commit messages without breaking a previous commit. What I can do is add another function just to do what `Execute` now doing => delete old suff and rename that function to `Execute`. I am not able to break this tightly coupled commit.

I wonder if the exit code stuff could be separated out... However, it might be the commit log that is missing details on the extent of the changes :) And why they are being done. Regarding the changes, it seems this switches from having some kind of 'shell server' which can be fed commands to execute, and returns the result, to direct invocation of the command, at the cost of starting the shell once per command?
I think you did that because of a bug which you were experiencing in CI, but the commit log says nothing about it, while it's not far from being the most important thing to have for future reference (why is this being done?).

praveenkumar commented:
> I wonder if the exit code stuff could be separated out.

@cfergeau Updated the commit logs.  If I do that then I have to keep the old methods to get the exitcode which also doesn't work in case of `podman build` command.

cfergeau commented on features/testsuite.feature:
```diff
@@ -1,13 +1,15 @@
      Go is a tool for managing Go source code.
      """
 
+  @linux @darwin
```
Why? The commit log does not say anything about this.

cfergeau commented on appveyor.yml:
```diff
@@ -4,4 +4,4 @@ environment:
 stack: go 1.14
 test_script:
-  - go test
+  - go test --godog.tags=windows
```
Should this be squashed with the commit which added the @linux,@windows, ... tags?

praveenkumar replied:
This is specific to CI so I have as separate commit.

cfergeau commented on features/testsuite.feature:
```diff
@@ -64,7 +64,7 @@ Quentin check whether their testsuite works properly.
 
-   # Config 
+   # Config
 
```
These changes are unrelated.

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
-func StartHostShellInstance(shellName string) error {
-	return shell.Start(shellName)
+func SetUpHostShellInstance(shellName string) error {
+	return shell.SetUp(shellName)
```
Maybe `Setup` rather than `SetUp`?

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
-		return err
-	}
 
 	fmt.Printf("The %v instance has been started and will be used for testing.\n", shell.name)
```
This message can probably be dropped. It at least needs to be reworded, as nothing is started anymore.

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
-	if shell.instance == nil {
+	if shell == nil {
 		return errors.New("shell instance is not started")
 	}
```
I don't think `shell` can be `nil`. `shell.name` could be empty though. And the `started` reference is obsolete.

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
+	cmd := exec.Command(shell.name, shell.startArgument...)
+	cmd.Stderr = &shell.errbuf
+	cmd.Stdout = &shell.outbuf
+	cmd.Stdin = strings.NewReader(command + "\n")
```
The log says
> We still create the input Pipe and pass the command in a goroutine but just after that
executing this stdinPipe closed 

I think this is no longer  correct?

cfergeau commented on testsuite/shell.go:
```diff
@@ -128,103 +85,57 @@ func (shell *ShellInstance) ConfigureTypeOfShell(shellName string) {
+		case errors.As(err, &pathError):
+			return fmt.Errorf("os.PathError: %w", err)
+		default:
+			return fmt.Errorf("something went worng with %s command: %w", command, err)
```
'wrong'
